### ISSUE TITLE
Fields with `Optional` type are cumbersome to initialize

### ISSUE THREAD

**Issue by Grannath** [immutables/immutables#435]
When using fields with an Optional type, the builder accepts a value in two ways:

```Java
class Builder {
    public Builder optionalField(Object optionalField);

    public Builder optionalField(Optional<Object> optionalField);
}
```

The first would be the way to interface with an API that worked with `null` instead of `Optional`. However, it throws a `NullPointerException` when handed a `null` reference.

This was probably a conscious decision as it makes sure unwanted empty values are caught early. But on the other hand it means that although I explicitly created the field as `Optional`, I have to manually work around `null`.

I believe that the much more common case is that `null` should be mapped to an empty `Optional`. After all, the type already indicates that nothing is a common value for the field. A NPE is right for a field that has not been explicitly marked as possibly empty. This is the way "normal" fields are handled, depending on an `@Nullable` annotation. But `Optional` fields break this pattern.

It is not a common rule that although a field is of an `Optional` type, an empty value has to be explicitly prevented. If such a guarantee has to hold, it should be explicit in the code. Simply passing a reference and trusting that although it will be wrapped in an `Optional` it will never be empty is far from obvious.

In order to not break backwards compatibility I propose a config value to switch the behavior for the generated builder. The default would be as it is now. Setting the value to the non-default would use `Optional.ofNullable()` instead of `Optional.of()` and thus allow empty values by passing null.

Alternatively a third method could be generated.

```Java
class Builder {
    public Builder optionalField(Object optionalField);

    public Builder optionalFieldNullable(Object optionalField);

    public Builder optionalField(Optional<Object> optionalField);
}
```

But I consider this solution inferior as it further clutters the API.

I would like to get some feedback, preferably from other users of this feature on which behavior they expected.

**Comment by elucash** [comment]
Hi Johannes. You can read some previous threads on this #245 #152
TL;DR;

There's style `Value.Style(optionalAcceptNullable=true)` for this.
In addition to that, new experimental "encodings" functionality #363 could possibly allow custom ways to code any special kind of attribute, including redefining how optionals are supported.
Thank you!

**Comment by Grannath** [comment]
Sorry, I did not see that style option. But in my defense, it should probably be mentioned in the documentation. Anyway, I'll close it then.

**Comment by elucash** [comment]
@Grannath thank you for your input! Your position which is acknowledged and respected. Maybe it's not mentioned in the docs, because of our a tendency to keep people away of equating null to optional empty (see https://github.com/immutables/immutables/issues/245#issuecomment-176305649), but it is fair that we should make a note of it in docs.
